src/components/PizzaList.js

```diff
@@ -0,0 +1,108 @@
+  const user = useSelector(selectUser);
+  const pizzas = useSelector(selectPizzas);
+  const dispatch = useDispatch();
+  const [favorites, setFavorites] = useState(null);
+
+  useEffect(() => {
+    const allIngredients = {};
+    pizzas.forEach((pizza) => {
+      pizza.ingredients?.forEach((ingredient) => {
+        allIngredients[ingredient] = false;
+      });
+    });
+    setFavorites(allIngredients);
+  }, [pizzas]);
+
+  return (
+    <div id="pizzaList">
+      <h1>Pizza Explorer</h1>
+      <p>
+        Welcome back, <strong>{user.name}</strong>! Your favorite pizzas:
+      </p>
+      <p>total number of known pizzas: {pizzas.length}</p>
+      <div id="pizzaContainer">
+        {favorites &&
+          pizzas
+            .filter((pizza) => {
+              let include = true;
+              Object.keys(favorites).forEach((ingredient) => {
+                if (favorites[ingredient] === true) {
+                  if (
+                    !pizza.ingredients ||
+                    !pizza.ingredients.includes(ingredient)
+                  ) {
+                    include = false;
+                    return;
+                  }
+                }
+              });
+              return include;
+            })
+            .map((pizza) => {
+              return (
+                <div className="pizzaCard" key={pizza.id}>
+                  {pizza.url && <img src={pizza.url} alt={pizza.name}></img>}
+                  <div className="favorite">
+                    <button
+                      className={
+                        user.favorites.includes(pizza.id) ? "fav" : "noFav"
+                      }
+                      onClick={() => {
+                        dispatch({
+                          type: "TOGGLE_FAVORITE_PIZZA",
+                          payload: pizza.id,
+                        });
+                      }}
+                    >
+                      {user.favorites.includes(pizza.id) ? "♥" : "♡"}
+                    </button>
+                  </div>
+                  <div className="pizzaText">
+                    <p>
+                      <strong>{pizza.name}:</strong>
+                      <br></br>
+                      {pizza.description} <i>Bought {pizza.bought} times.</i>
+                      <br></br>
+                      <br></br>
+                      {pizza.ingredients && pizza.ingredients.join(", ")}
+                    </p>
+                  </div>
+                </div>
+              );
+            })}
+      </div>
+      <h2>Filter by ingrendients:</h2>
+      {favorites &&
+        Object.keys(favorites).map((ingredient) => {
+          return (
+            <button
+              key={ingredient}
+              className={favorites[ingredient] ? "on" : "off"}
+              onClick={() => {
+                const allIngredientsCopy = {
+                  ...favorites,
+                };
+                allIngredientsCopy[ingredient] = !favorites[ingredient];
```

Comment:
Now you're building the buttons from the favourites, which makes sense because you already did the hard work of building it up.
For this I would use ```.flatMap``` in this way:
```
const allIng = pizzas.flatMap(p => p.ingredients) // this gives you an array with all the ingredients BUT with duplicates. 
const uniqueIngredients = [...new Set(allIng)] // The set constructor only keeps the unique values, then we turn it back into an array
```